src/rhsm/connection.py

```diff
@@ -762,54 +860,56 @@ def _set_accept_language_in_header(self):
         if lc:
             self.headers["Accept-Language"] = lc.lower().replace("_", "-").split(".", 1)[0]
 
-    # FIXME: can method be empty?
-    def _request(self, request_type, method, info=None, headers=None, cert_key_pairs=None):
-        handler = self.apihandler + method
-
-        # We try to import it here to get fresh value, because rhsm.service can receive
-        # several D-BUS API calls with different locale argument (every request have to have
-        # different locale)
-        self._set_accept_language_in_header()
-
-        # Load certificates from cert dir if specified
-        if cert_key_pairs is None or len(cert_key_pairs) == 0:
-            cert_key_pairs = self._get_cert_key_list()
-
-        if (
-            headers is not None
-            and "Content-type" in headers
-            and headers["Content-type"] == "application/x-www-form-urlencoded"
-        ):
-            body = urlencode(info).encode("utf-8")
-        elif info is not None:
-            body = json.dumps(info, default=json.encode)
-        else:
-            body = None
-
-        log.debug("Making request: %s %s" % (request_type, handler))
-
-        if self.user_agent:
-            self.headers["User-Agent"] = self.user_agent
-
-        final_headers = self.headers.copy()
-        if body is None:
-            final_headers["Content-Length"] = "0"
-        if headers:
-            final_headers.update(headers)
-
-        self._print_debug_info_about_request(request_type, handler, final_headers, body)
-
+    @staticmethod
+    def parse_keep_alive_header(keep_alive_header: str) -> tuple:
+        """
+        Try to parse 'Keep-Alive' header received from candlepin server
+        :param keep_alive_header: string with value of the header
+        :return: Tuple containing connection timeout and maximal number of requests
+        """
+        keep_alive_timeout = None
+        max_requests_num = None
+        # Regular expression pattern represents: key=number
+        pattern = re.compile(r"^(.*)=(\d+)$")
+
+        items = keep_alive_header.split()
+
+        for item in items:
+            search_result = pattern.search(item)
+            if search_result is not None:
+                key, value = search_result.groups()
+                # Timeout of connection using keep-alive
+                if key == "timeout":
+                    keep_alive_timeout = int(search_result.groups()[1])
+                # Maximal number of request on one connection
+                if key == "max":
+                    max_requests_num = int(search_result.groups()[1])
```

Comment:
this can be simplified a bit without a regexp:
```diff
--- a/src/rhsm/connection.py
+++ b/src/rhsm/connection.py
@@ -869,21 +869,19 @@ class BaseRestLib(object):
         """
         keep_alive_timeout = None
         max_requests_num = None
-        # Regular expression pattern represents: key=number
-        pattern = re.compile(r"^(.*)=(\d+)$")
 
         items = keep_alive_header.split()
 
         for item in items:
-            search_result = pattern.search(item)
-            if search_result is not None:
-                key, value = search_result.groups()
+            parts = item.split("=", maxsplit=2)
+            if len(parts) == 2:
+                key, value = parts
                 # Timeout of connection using keep-alive
                 if key == "timeout":
-                    keep_alive_timeout = int(search_result.groups()[1])
+                    keep_alive_timeout = safe_int(value)
                 # Maximal number of request on one connection
-                if key == "max":
-                    max_requests_num = int(search_result.groups()[1])
+                elif key == "max":
+                    max_requests_num = safe_int(value)
 
         return keep_alive_timeout, max_requests_num
 
```
also please use a `if`/`elif`/etc chain, so it is a minor optimization

Reply:
I think that I don't like it, because it is not reliable. See:
```python
s1 = "foo = 123 = 546"
s1.split("=", maxsplit=1)
```
Produces following result: `['foo ', ' 123 = 546']`. There cannot be any empty spaces around `=`, but your proposal of code does not have any problem with that. Of course you can create more reliable variant, but I think that regular expression is still better option.

I would not use regular expression on server side, but it is client application and we are not going to parse many HTTP headers per second.

Reply:
> BTW: Your proposal does not have problem with following argument: `timeout=123=456`. It will produce timeout with string `"123=465"` and it could cause troubles later. I still believe that regular expression is more reliable in this case.

I'm not sure I understand: even in my proposal there is the conversion to int, and `"123=465"` definitely makes `int()` raise ValueError.

I checked my proposal with the tests you added, and they work fine.

Reply:
> > BTW: Your proposal does not have problem with following argument: `timeout=123=456`. It will produce timeout with string `"123=465"` and it could cause troubles later. I still believe that regular expression is more reliable in this case.
> 
> I'm not sure I understand: even in my proposal there is the conversion to int, and `"123=465"` definitely makes `int()` raise ValueError.

I'm sorry for confusion. I tested only small snippet of your proposal. `safe_int()` will not raise exception, but it will return `safe_value` which is default `None` in this case. It is probably fine, but floats and negative values should not be IMHO allowed according RFC.

I added more test cases.

Reply:
> It is probably fine, but floats and negative values should not be IMHO allowed according RFC.

`int()` already rejects floats. Regarding negative values, it can be an easy addition:
```python
def safe_positive_int(value, safe_value=None):
    num = safe_int(value, safe_value)
    if num is not None and num <= 0:
        return safe_value
    return num
````

Reply:
I just don't like what you are proposing. Please stop enforce me to rewrite my code in different way. There is nothing wrong with regular expression.

Reply:
Jirka, no need to be that drastic nor absolute :) We are having a polite conversation on the possible implementation of this, and I'm explaining my ideas with reasons, examples and code snippets. There is noone "enforcing you" in any way, I'm sorry if you feel that way.

I said earlier why I don't think it's ideal to use a regexp: while it works, it's too slow, and can be done even without.

Reply:
I know that it is slow, but we talk about client (not server). We do not parse 10000 requests per second. We parse one response per second, when everything goes smoothly.